CKS Enhancements docs (#458)
apache/cloudstack-documentation: source/plugins/cloudstack-kubernetes-service.rst

The Kubernetes Service plugin adds Kubernetes integration to CloudStack. The plugin is disabled by default and an admin can enable it using a Global Setting. It enables users to run containerized services using Kubernetes clusters.

> With CoreOS having reached EOL, from 4.16 the Kubernetes Service Plugin will use the existing SystemVM Template by default for deploying kubernetes clusters. For installation of Kubernetes binaries on the cluster nodes, a binaries ISO is used for each Kubernetes version to be made available via CloudStack. This allows faster, offline installation of Kubernetes binaries and docker images along with support for adding multiple versions of Kubernetes for upgrades and running different clusters.

Reviewer: ```suggestion
With CoreOS having reached EOL, from 4.16 on the Kubernetes Service Plugin will use
the existing SystemVM Template by default for deploying kubernetes clusters. For
installation of Kubernetes binaries on the cluster nodes, a binaries ISO is used for each
Kubernetes version to be made available via CloudStack. This allows faster, offline
installation of Kubernetes binaries and docker images along with support for adding
multiple versions of Kubernetes for upgrades and running different clusters.
```

.. note::

>    Since version 4.21.0 users can choose different templates for different types of nodes (worker, control, etcd nodes) for deploying Kubernetes clusters. These templates must be previously registered selecting the 'For CKS' option.

Reviewer: ```suggestion
   Since version 4.21.0 users can choose different templates for different types of nodes
   (worker, control, etcd nodes) for deploying Kubernetes clusters. These templates must
   be previously registered selecting the 'For CKS' option.
```

|cks-add-version-form.png|

.. note::

>    Since 4.21.0 it is possible to deploy separate dedicated etcd nodes. This requires the Kubernetes ISO contains the etcd binaries.

Reviewer: ```suggestion
   Since 4.21.0 it is possible to deploy separate dedicated etcd nodes. This requires
   the Kubernetes ISO to contains the etcd binaries.
```

|cks-create-cluster-form.png|

> Since 4.21.0, users will be provided with an additional section displayed when toggling the option: 'Show Advanced Settings'. On this section, users can select templates and service offerings for:

Reviewer: ```suggestion
Since 4.21.0, users will be provided with an additional section displayed when toggling
the option: 'Show Advanced Settings'. On this section, users can select templates and
service offerings for:
```

   These operations are only supported for an ExternalManaged Kubernetes Cluster

.. _flexible-kubernetes-clusters:

Flexible Kubernetes Clusters
----------------------------

Since 4.21.0, CloudStack introduces many enhancements to Kubernetes Clusters allowing users to:

- Select the Hypervisor hype for the Kubernetes Cluster nodes
- Specify different templates and/or service offerings for different types of Kubernetes Clusters nodes

> - Use CKS-ready custom and non-ready templates for Kubernetes cluster nodes

Author: Meant to say that whatever the template you register as 'For CKS' it can be selected for the CKS cluster node, even if it doesn't have the required packages (non-ready) as CS will verify if they have it

- Set the template specific values as usual for template registration.

- Mark the option 'For CKS'. This ensures the template is considered as an option for Kubernetes cluster nodes on the Advanced Settings section for clusters creation.

Separate etcd nodes from control nodes
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

By default, the number of etcd nodes in a CKS cluster is 0, the etcd service is included on the control nodes. If the number of etcd nodes passed on the CKS cluster creation is at least 1, then CloudStack will dedicate nodes only to the etcd service, separating them from the control nodes.

To use separate etcd nodes, it is required to build and register a CKS ISO version containing the etcd service as explained on: :ref:`kubernetes-supported-versions`

Add an external VM Instance as a worker node to a Kubernetes cluster
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Requirements for a VM Instance to be added as worker node to a Kubernetes cluster:

- At least 8GB ROOT disk size, 2 CPU cores and 2GB RAM

- The VM Instance must have a NIC on the Kubernetes cluster network

> - The Management Server’s SSH Public key must be added at the cloud user’s authorized_keys file at `~/.ssh/authorized_keys`.

Author: Sure, will do

- The process timeout is set by the setting: `cloud.kubernetes.cluster.remove.node.timeout`.

Dedicate specific hosts/clusters to a specific domain for CKS cluster deployment
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Administrators are able to dedicate hosts to a domain or account. CloudStack will take the host dedication into consideration when deploying Kubernetes clusters.

- When there are no hosts dedicated to the domain/account the user belongs, then the nodes will be deployed on any host.

- When there are hosts dedicated to the domain/account the user belongs, then the nodes will be deployed on the dedicated hosts.

   .. note::
      By design the hosts dedication does not consider the deployment of system VMs on the dedicated hosts (SSVM, CPVM and Virtual Routers). In case the Kubernetes cluster is created on an unimplemented network then the Virtual Router of the network will not be deployed on the dedicated hosts.

Use diverse CNI plugins (Calico, Cilium, etc)
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> A CNI framework has also been added which provides end users the flexibility to use the CNI plugin of their choice. The CNI framework internally leverages the managed Userdata feature provided by CloudStack.

Reviewer 2: ```suggestion
A CNI framework has also been added which provides end users the flexibility to use the CNI plugin of their choice. The CNI framework internally leverages the managed User data feature provided by CloudStack.
```

> Sample Calico CNI configuration data used which is appended to the existing Kubernetes control node userdata is:

Reviewer 2: ```suggestion
Sample Calico CNI configuration data used which is appended to the existing Kubernetes control node user data is:
```
